src/compile/data/flatten.ts

```diff
@@ -0,0 +1,48 @@
+import {FlattenTransform} from '../../transform';
+import {duplicate} from '../../util';
+// TODO Use vega-typings
+import {VgFlattenTransform} from '../../vega.schema';
+import {DataFlowNode} from './dataflow';
+
+/**
+ * A class for flatten transform nodes
+ */
+export class FlattenTransformNode extends DataFlowNode {
+  public clone() {
+    return new FlattenTransformNode(this.parent, duplicate(this.transform));
+  }
+
+  constructor(parent: DataFlowNode, private transform: FlattenTransform) {
+    super(parent);
+  }
+
+  public producedFields() {
+    const out = {};
+    this.transform.flatten.forEach(field => {
+      out[this.getDefaultName(field)] = true;
+    });
+    return out;
+  }
+
+  private getDefaultName(field: string): string {
+    const index = this.transform.flatten.indexOf(field);
+    // Returns the "as" entry corresponding to field if it exists else returns field
+    return (this.transform.as === undefined || this.transform.as[index] === undefined) ? field : this.transform.as[index];
```

Comment:
Right now this follows Vega Defaults. Not sure whether to change this and introduce different behavior when `as` is not specified.

Reply:
Can't we just pass nothing and let Vega handle the default case?

Reply:
@kanitw mentioned in slack that we need to set `as`, so that `producedFields` gives the correct output.  But here, since the default is the `field` value, I think passing nothing might be ok.

Reply:
Try it out and I'd always prefer to output less.

Reply:
You can output `as=undefined` if you modify `producedFields` to output `field` name instead if `as` is undefined

Reply:
I think this method is pretty unnecessary.

Reply:
> Try it out and I'd always prefer to output less.

I just check the rest of transform are all explicit though, so doing so would be quite inconsistent.  So let's make them explicit for consistency for now. 

If @domoritz doesn't like it, he can refactor all of the transforms later.